006-tls-encrypt-the-kafka-connect-rest-api.md

```diff
@@ -0,0 +1,98 @@
+rest.advertised.listener: https
+rest.advertised.port: 8443
+listeners.https.ssl.client.auth: none
+listeners.https.ssl.truststore.location: /tmp/kafka/kafka-connect-rest.truststore.p12
+listeners.https.ssl.truststore.password: ***generated password***
+listeners.https.ssl.truststore.type: PKCS12
+listeners.https.ssl.keystore.location: /tmp/kafka/kafka-connect-rest.keystore.p12
+listeners.https.ssl.keystore.password: ***generated password***
+listeners.https.ssl.keystore.type: PKCS12
+```
+
+To communicate with the encrypted Connect REST API listener, the `KafkaMirrorMaker2`, `KafkaConnect` and `KafkaConnector` operators require modifications to the `AbstractConnectOperator` to apply the TLS certs as keystore credentials to a WebClient.
+
+### Enabling a plain HTTP REST API listener
+
+A new field is added to the `KafkaConnect` and `KafkaMirrorMaker2` specs named `restListeners`, which allows users to configure an additional unencrypted HTTP listener on port 8083 in the Kafka Connect configuration.
+This enables users to access the REST API without TLS if required.
+The `KafkaConnector`, `KafkaConnect` and `KafkaMirrorMaker2` operators will continue to use the TLS encrypted listener when an unencrypted HTTP listener is enabled.
+
+The `restListeners` field is designed for potential future extension, and is a list of items that contain a `protocol` field.
+For this proposal, the `protocol` field only supports the value `http`. One (or more) items in the `restListeners` list will add the unencrypted HTTP listener on port 8083.
+
+For example, the following CR will enable the HTTP listener on port 8083:
+```
+apiVersion: kafka.strimzi.io/v1beta1
+kind: KafkaConnect
+spec:
+  restListeners:
+  - protocol: http
+  ...
+```
+
+The schema for `restListeners` is as follows:
+```
+openAPIV3Schema:
+  type: object
+  properties:
+    spec:
+      type: object
+      properties:
+        restListeners:
+          type: array
+          items:
+            type: object
+            properties:
+              protocol:
+                type: string
+                enum:
+                - http
+                description: The protocol of the REST listener.
+            required:
+            - protocol
+          description: List of additional REST listeners.
+        ...
+```
+
+#### Annotation alternative to enable a plain HTTP REST API listener
+
+Alternatively, a new annotation `strimzi.io/enable-plain-rest-listener` could be added, to avoid changing the spec.
+Setting the `strimzi.io/enable-plain-rest-listener` annotation value to `true` would add the additional unencrypted HTTP listener on port 8083.
+
+
+### Certificate renewals
+
+The self-signed TLS certificate will have a 1-year expiration.
+One day before the self-signed TLS certificate expires, the operator will automatically renew the certificate, replace the old self-signed certificate, and restart the pods in a controlled manner to ensure the new certificates are in use by Kafka Connect.
```

Comment:
I assume this is done on the periodic reconcile? Or a reconcile triggered due to the certificate renewal period?

Reply:
Yes, on the periodic reconcile. We could also add in [maintenance time windows](https://strimzi.io/docs/operators/latest/using.html#assembly-maintenance-time-windows-deployment-configuration-kafka), and/or support the `strimzi.io/force-renew` and `strimzi.io/force-replace` annotations to allow users to force new certs and keys to be created. What do you think?

Looking through the Kafka operator code I see that the default renewal period is 30 days before the cert expiry (which has a default of 365 days) - I will update the proposal to change 'One day before..' to '30 days before...'  so that the default periods match. Do you think these certificates need similar fine-grained control that users get in the Kafka `spec.clusterCa` fields?

Reply:
I'd like to hear the other maintainers opinions on this, but this sounds reasonable to me.